File: sdk/src/canonical.ts
/**
 * DisputeVault canonical JSON (PROTOCOL.md §2, §4, §7).
 *
 * Deterministic serialization shared by frontend, gateway, and (as a
 * behavioral spec) the Go worker:
 *   - fixed field order (schema order, never alphabetical)
 *   - UTF-8, no BOM, no whitespace
 *   - integers: plain decimal ASCII, no leading '+', no leading zeros, no
 *     exponents, no fraction
 *   - strings: minimal escaping (", \, control chars as \u00XX); non-ASCII
 *     emitted raw as UTF-8
 *   - hash fields: exactly 32 lowercase hex chars
 *   - unknown fields rejected, never ignored
 *   - size budgets enforced in BYTES (UTF-8), never truncated
 */
import {
  DISPUTE_TYPES,
  EVIDENCE_SUMMARY_MAX_CHARS,
  VERDICT_DECISIONS,
  VERDICT_POLICY_VERSION_MAX_CHARS,
  VERDICT_REASON_CODE_MAX_CHARS,
  WIRE,
  type DisputeType,
  type EvidenceV1,
  type VerdictDecision,
  type VerdictV1,
} from './schema.js';

/** Thrown for every protocol violation; message names the violated rule. */
export class ProtocolError extends Error {
  constructor(message: string) {
    super(`disputevault: ${message}`);
    this.name = 'ProtocolError';
  }
}

const encoder = new TextEncoder();

export const byteLength = (s: string): number => encoder.encode(s).length;

const HEX32 = /^[0-9a-f]{32}$/;
const REASON_CODE = /^[A-Z0-9_]{1,32}$/;
const POLICY_VERSION = /^[a-z0-9-]{1,32}$/;

function isPlainObject(x: unknown): x is Record<string, unknown> {
  return typeof x === 'object' && x !== null && !Array.isArray(x);
}

/** Minimal JSON string escaping per PROTOCOL.md §2 rule 5. */
export function escapeJsonString(s: string): string {
  let out = '"';
  for (const ch of s) {
    // iterate by code point so surrogate pairs survive intact
    const code = ch.codePointAt(0)!;
    if (ch === '"') out += '\\"';
    else if (ch === '\\') out += '\\\\';
    else if (code < 0x20) {
      out += '\\u' + code.toString(16).padStart(4, '0');
    } else {
      out += ch; // non-ASCII stays raw (UTF-8), never \uXXXX
    }
  }
  return out + '"';
}

/** Plain decimal ASCII for a non-negative safe integer (§2 rule 4). */
function integerToJson(n: number | bigint): string {
  const value = typeof n === 'bigint' ? n : Math.trunc(n);
  if (value < 0) throw new ProtocolError(`negative integer not allowed: ${n}`);
  if (typeof n === 'number' && !Number.isSafeInteger(n)) {
    throw new ProtocolError(`integer exceeds the safe range: ${n}`);
  }
  return value.toString(10);
}

/**
 * Serialize an already-validated payload to canonical JSON, using the fixed
 * field order of its schema. Field values are emitted raw; validation is the
 * caller's job (validateEvidence / validateVerdict).
 */
function serializeOrdered(fields: Array<[string, string]>): string {
  return '{' + fields.map(([k, v]) => `${escapeJsonString(k)}:${v}`).join(',') + '}';
}

// ---------------------------------------------------------------------------
// EvidenceV1 (§4)
// ---------------------------------------------------------------------------

/** Validate an EvidenceV1 payload against PROTOCOL.md §4. Throws ProtocolError. */
export function validateEvidence(payload: unknown): EvidenceV1 {
  if (!isPlainObject(payload)) {
    throw new ProtocolError('evidence payload must be a JSON object');
  }
  const allowed = ['v', 'bounty_id', 'claimant_id_hash', 'type', 'evidence_ref_hash', 'summary'];
  for (const key of Object.keys(payload)) {
    if (!allowed.includes(key)) {
      throw new ProtocolError(`evidence payload has unknown field: ${key}`);
    }
  }
  for (const key of allowed) {
    if (!(key in payload)) {
      throw new ProtocolError(`evidence payload is missing field: ${key}`);
    }
  }
  const p = payload as Record<string, unknown>;

  if (p.v !== 1) throw new ProtocolError(`evidence v must be 1, got ${JSON.stringify(p.v)}`);
  if (typeof p.bounty_id !== 'number' || !Number.isInteger(p.bounty_id) || p.bounty_id < 0 || p.bounty_id > 0xffffffff) {
    throw new ProtocolError(`evidence bounty_id must be an integer in 0..4294967295, got ${JSON.stringify(p.bounty_id)}`);
  }
  if (typeof p.claimant_id_hash !== 'string' || !HEX32.test(p.claimant_id_hash)) {
    throw new ProtocolError('evidence claimant_id_hash must be 32 lowercase hex chars');
  }
  if (typeof p.type !== 'number' || !DISPUTE_TYPES.includes(p.type as DisputeType)) {
    throw new ProtocolError(`evidence type must be one of ${DISPUTE_TYPES.join(', ')}`);
  }
  if (typeof p.evidence_ref_hash !== 'string' || !HEX32.test(p.evidence_ref_hash)) {
    throw new ProtocolError('evidence evidence_ref_hash must be 32 lowercase hex chars');
  }
  if (typeof p.summary !== 'string') {
    throw new ProtocolError('evidence summary must be a string');
  }
  if ([...p.summary].length > EVIDENCE_SUMMARY_MAX_CHARS) {
    throw new ProtocolError(`evidence summary exceeds ${EVIDENCE_SUMMARY_MAX_CHARS} characters`);
  }
  return {
    v: 1,
    bounty_id: p.bounty_id,
    claimant_id_hash: p.claimant_id_hash,
    type: p.type as DisputeType,
    evidence_ref_hash: p.evidence_ref_hash,
    summary: p.summary,
  };
}

/**
 * Canonicalize an EvidenceV1 payload (§2 + §4). Throws ProtocolError if the
 * payload violates the schema or the canonical JSON exceeds the 250-byte
 * evidence budget - nothing is ever truncated.
 */
export function canonicalizeEvidence(payload: unknown): string {
  const e = validateEvidence(payload);
  const json = serializeOrdered([
    ['v', integerToJson(e.v)],
    ['bounty_id', integerToJson(e.bounty_id)],
    ['claimant_id_hash', escapeJsonString(e.claimant_id_hash)],
    ['type', integerToJson(e.type)],
    ['evidence_ref_hash', escapeJsonString(e.evidence_ref_hash)],
    ['summary', escapeJsonString(e.summary)],
  ]);
  const bytes = byteLength(json);
  if (bytes > WIRE.evidenceMaxJsonBytes) {
    throw new ProtocolError(
      `evidence canonical JSON is ${bytes} bytes; budget is ${WIRE.evidenceMaxJsonBytes} (PROTOCOL §3) - rejected, never truncated`,
    );
  }
  return json;
}

// ---------------------------------------------------------------------------
// VerdictV1 (§7)
// ---------------------------------------------------------------------------

const MAX_DISPUTE_ID = (1n << 64n) - 1n;

/** Validate a VerdictV1 payload against PROTOCOL.md §7. Throws ProtocolError. */
export function validateVerdict(payload: unknown): VerdictV1 {
  if (!isPlainObject(payload)) {
    throw new ProtocolError('verdict payload must be a JSON object');
  }
  const allowed = ['v', 'dispute_id', 'decision', 'confidence', 'reason_code', 'policy_version'];
  for (const key of Object.keys(payload)) {
    if (!allowed.includes(key)) {
      throw new ProtocolError(`verdict payload has unknown field: ${key}`);
    }
  }
  for (const key of allowed) {
    if (!(key in payload)) {
      throw new ProtocolError(`verdict payload is missing field: ${key}`);
    }
  }
  const p = payload as Record<string, unknown>;

  if (p.v !== 1) throw new ProtocolError(`verdict v must be 1, got ${JSON.stringify(p.v)}`);
  const id = p.dispute_id;
  const idBig =
    typeof id === 'bigint' ? id : typeof id === 'number' && Number.isInteger(id) && id >= 0 ? BigInt(id) : undefined;
  if (idBig === undefined || idBig < 0n || idBig > MAX_DISPUTE_ID) {
    throw new ProtocolError(`verdict dispute_id must be an integer in 0..2^64-1, got ${String(id)}`);
  }
  if (typeof p.decision !== 'string' || !VERDICT_DECISIONS.includes(p.decision as VerdictDecision)) {
    throw new ProtocolError(`verdict decision must be one of ${VERDICT_DECISIONS.join(' | ')}`);
  }
  if (typeof p.confidence !== 'number' || !Number.isInteger(p.confidence) || p.confidence < 0 || p.confidence > 100) {
    throw new ProtocolError(`verdict confidence must be an integer in 0..100, got ${JSON.stringify(p.confidence)}`);
  }
  if (typeof p.reason_code !== 'string' || !REASON_CODE.test(p.reason_code) || p.reason_code.length > VERDICT_REASON_CODE_MAX_CHARS) {
    throw new ProtocolError('verdict reason_code must match [A-Z0-9_] with 1..32 chars');
  }
  if (typeof p.policy_version !== 'string' || !POLICY_VERSION.test(p.policy_version) || p.policy_version.length > VERDICT_POLICY_VERSION_MAX_CHARS) {
    throw new ProtocolError('verdict policy_version must match [a-z0-9-] with 1..32 chars');
  }
  return {
    v: 1,
    dispute_id: idBig,
    decision: p.decision as VerdictDecision,
    confidence: p.confidence,
    reason_code: p.reason_code,
    policy_version: p.policy_version,
  };
}

/**
 * Canonicalize a VerdictV1 payload (§2 + §7). Throws ProtocolError if the
 * payload violates the schema or the canonical JSON exceeds the 186-byte
 * verdict budget - nothing is ever truncated.
 */
export function canonicalizeVerdict(payload: unknown): string {
  const v = validateVerdict(payload);
  const json = serializeOrdered([
    ['v', integerToJson(v.v)],
    ['dispute_id', integerToJson(v.dispute_id)],
    ['decision', escapeJsonString(v.decision)],
    ['confidence', integerToJson(v.confidence)],
    ['reason_code', escapeJsonString(v.reason_code)],
    ['policy_version', escapeJsonString(v.policy_version)],
  ]);
  const bytes = byteLength(json);
  if (bytes > WIRE.verdictMaxJsonBytes) {
    throw new ProtocolError(
      `verdict canonical JSON is ${bytes} bytes; budget is ${WIRE.verdictMaxJsonBytes} (PROTOCOL §3) - rejected, never truncated`,
    );
  }
  return json;
}

/** Convenience: parse + validate + canonicalize from raw JSON text. */
export function canonicalizeEvidenceJson(text: string): string {
  let parsed: unknown;
  try {
    parsed = JSON.parse(text);
  } catch {
    throw new ProtocolError('evidence payload is not valid JSON');
  }
  return canonicalizeEvidence(parsed);
}

/** Convenience: parse + validate + canonicalize from raw JSON text. */
export function canonicalizeVerdictJson(text: string): string {
  let parsed: unknown;
  try {
    parsed = JSON.parse(text);
  } catch {
    throw new ProtocolError('verdict payload is not valid JSON');
  }
  return canonicalizeVerdict(parsed);
}

File: worker/internal/canonical/canonical.go
// Package canonical implements DisputeVault EvidenceV1/VerdictV1 payloads
// and deterministic canonical JSON (PROTOCOL.md §2, §4, §7), mirroring
// sdk/src/canonical.ts and sdk/src/schema.ts byte-for-byte.
//
// Canonical rules (§2):
//   - fixed field order (schema order, never alphabetical)
//   - UTF-8, no BOM, no whitespace
//   - integers: plain decimal ASCII
//   - strings: minimal escaping (", \, control chars as \u00XX); non-ASCII
//     raw UTF-8, never \uXXXX
//   - hash fields: exactly 32 lowercase hex chars
//   - size budgets enforced in BYTES, never truncated
//
// Go's encoding/json is deliberately NOT used for serialization: it escapes
// non-ASCII by default, HTML-escapes <>&, and orders map fields
// alphabetically. The encoders here write the exact bytes the SDK writes.
package canonical

import (
	"errors"
	"fmt"
	"regexp"
	"strings"
)

// ProtocolError is returned for every protocol violation; the message names
// the violated rule (mirrors sdk ProtocolError).
type ProtocolError struct{ Msg string }

func (e *ProtocolError) Error() string { return "disputevault: " + e.Msg }

// NewProtocolError wraps a message into a *ProtocolError for use by sibling
// packages (wire, commitment).
func NewProtocolError(format string, args ...any) error {
	return &ProtocolError{Msg: fmt.Sprintf(format, args...)}
}

func errf(format string, args ...any) error {
	return &ProtocolError{Msg: fmt.Sprintf(format, args...)}
}

// Wire budgets (PROTOCOL.md §3).
const (
	EvidenceWireSize = 256
	EvidenceMaxJSON  = 250
	VerdictWireSize  = 192
	VerdictMaxJSON   = 186
)

// Field limits.
const (
	EvidenceSummaryMaxChars      = 64 // PROTOCOL §4 (characters, not bytes)
	VerdictReasonCodeMaxChars    = 32
	VerdictPolicyVersionMaxChars = 32
)

// Dispute types (§4).
const (
	TypeRejectedValidWork  = 0
	TypePaymentNotReceived = 1
	TypeScopeDispute       = 2
	TypeOther              = 3
)

// Verdict decisions (§7).
const (
	DecisionClaimantWins     = "claimant_wins"
	DecisionPlatformWins     = "platform_wins"
	DecisionPartial          = "partial"
	DecisionNeedsHumanReview = "needs_human_review"
)

var (
	hex32         = regexp.MustCompile(`^[0-9a-f]{32}$`)
	reasonCode    = regexp.MustCompile(`^[A-Z0-9_]{1,32}$`)
	policyVersion = regexp.MustCompile(`^[a-z0-9-]{1,32}$`)
)

var validDecisions = map[string]bool{
	DecisionClaimantWins:     true,
	DecisionPlatformWins:     true,
	DecisionPartial:          true,
	DecisionNeedsHumanReview: true,
}

// EvidenceV1 mirrors PROTOCOL.md §4; the struct order documents the canonical
// field order but encoding is explicit in CanonicalJSON.
type EvidenceV1 struct {
	V               uint32 // must be 1
	BountyID        uint32 // 0..4294967295
	ClaimantIDHash  string // 32 lowercase hex chars
	Type            uint8  // 0..3
	EvidenceRefHash string // 32 lowercase hex chars
	Summary         string // <= 64 chars, must also fit the byte budget
}

// VerdictV1 mirrors PROTOCOL.md §7.
type VerdictV1 struct {
	V             uint32 // must be 1
	DisputeID     uint64 // 0..2^64-1; MUST equal the outer dispute id (§7)
	Decision      string // enum
	Confidence    uint8  // 0..100
	ReasonCode    string // [A-Z0-9_], 1..32
	PolicyVersion string // [a-z0-9-], 1..32
}

// EscapeJSONString applies minimal JSON escaping per §2 rule 5: only ",
// \, and control characters (as \u00XX) are escaped; non-ASCII stays raw.
// It iterates by Unicode code point (the SDK iterates code points too), so
// surrogate-pair semantics are identical.
func EscapeJSONString(s string) string {
	var b strings.Builder
	b.WriteByte('"')
	for _, r := range s {
		switch {
		case r == '"':
			b.WriteString(`\"`)
		case r == '\\':
			b.WriteString(`\\`)
		case r < 0x20:
			b.WriteString(fmt.Sprintf(`\u%04x`, r))
		default:
			b.WriteRune(r) // raw UTF-8
		}
	}
	b.WriteByte('"')
	return b.String()
}

// CanonicalJSON serializes a validated EvidenceV1 payload to canonical JSON.
// It returns (json, error): error if the payload violates §4 or the canonical
// JSON exceeds the 250-byte evidence budget (§3) - nothing is ever truncated.
func (e EvidenceV1) CanonicalJSON() (string, error) {
	if err := e.validate(); err != nil {
		return "", err
	}
	var b strings.Builder
	b.WriteString(`{"v":1,`)
	b.WriteString(`"bounty_id":` + itoa(uint64(e.BountyID)) + `,`)
	b.WriteString(`"claimant_id_hash":` + EscapeJSONString(e.ClaimantIDHash) + `,`)
	b.WriteString(`"type":` + itoa(uint64(e.Type)) + `,`)
	b.WriteString(`"evidence_ref_hash":` + EscapeJSONString(e.EvidenceRefHash) + `,`)
	b.WriteString(`"summary":` + EscapeJSONString(e.Summary))
	b.WriteByte('}')

	json := b.String()
	if n := len([]byte(json)); n > EvidenceMaxJSON {
		return "", errf("evidence canonical JSON is %d bytes; budget is %d (PROTOCOL §3) - rejected, never truncated", n, EvidenceMaxJSON)
	}
	return json, nil
}

func (e EvidenceV1) validate() error {
	if e.V != 1 {
		return errf("evidence v must be 1, got %d", e.V)
	}
	if !hex32.MatchString(e.ClaimantIDHash) {
		return errors.New("disputevault: evidence claimant_id_hash must be 32 lowercase hex chars")
	}
	if e.Type > 3 {
		return errf("evidence type must be one of 0, 1, 2, 3 (got %d)", e.Type)
	}
	if !hex32.MatchString(e.EvidenceRefHash) {
		return errors.New("disputevault: evidence evidence_ref_hash must be 32 lowercase hex chars")
	}
	if n := len([]rune(e.Summary)); n > EvidenceSummaryMaxChars {
		return errf("evidence summary exceeds %d characters", EvidenceSummaryMaxChars)
	}
	return nil
}

// CanonicalJSON serializes a validated VerdictV1 payload to canonical JSON.
// Error if the payload violates §7 or the canonical JSON exceeds the
// 186-byte verdict budget (§3).
func (v VerdictV1) CanonicalJSON() (string, error) {
	if err := v.validate(); err != nil {
		return "", err
	}
	var b strings.Builder
	b.WriteString(`{"v":1,`)
	b.WriteString(`"dispute_id":` + itoa(v.DisputeID) + `,`)
	b.WriteString(`"decision":` + EscapeJSONString(v.Decision) + `,`)
	b.WriteString(`"confidence":` + itoa(uint64(v.Confidence)) + `,`)
	b.WriteString(`"reason_code":` + EscapeJSONString(v.ReasonCode) + `,`)
	b.WriteString(`"policy_version":` + EscapeJSONString(v.PolicyVersion))
	b.WriteByte('}')

	json := b.String()
	if n := len([]byte(json)); n > VerdictMaxJSON {
		return "", errf("verdict canonical JSON is %d bytes; budget is %d (PROTOCOL §3) - rejected, never truncated", n, VerdictMaxJSON)
	}
	return json, nil
}

func (v VerdictV1) validate() error {
	if v.V != 1 {
		return errf("verdict v must be 1, got %d", v.V)
	}
	if !validDecisions[v.Decision] {
		return errors.New("disputevault: verdict decision must be one of claimant_wins | platform_wins | partial | needs_human_review")
	}
	if v.Confidence > 100 {
		return errf("verdict confidence must be an integer in 0..100, got %d", v.Confidence)
	}
	if !reasonCode.MatchString(v.ReasonCode) || len(v.ReasonCode) > VerdictReasonCodeMaxChars {
		return errors.New("disputevault: verdict reason_code must match [A-Z0-9_] with 1..32 chars")
	}
	if !policyVersion.MatchString(v.PolicyVersion) || len(v.PolicyVersion) > VerdictPolicyVersionMaxChars {
		return errors.New("disputevault: verdict policy_version must match [a-z0-9-] with 1..32 chars")
	}
	return nil
}

// itoa renders a non-negative integer as plain decimal ASCII (§2 rule 4).
func itoa(n uint64) string {
	return fmt.Sprintf("%d", n)
}
